settings.txt
# This is the settings file for FracGen
# FracGen is (c) Douglas Wolf, 2017


FORMAT = 1

POI =
{
    "4 leaf clover", 0, center, -1.749537608906249986, -0.000000732421874941, 0.000007614843750003
    "Julia Dragon", 1, center, -0.920062500000000116, -0.280328125000000095, 0.236031250000000137
    "Lightning", 0, center, 0.409423828125000333, -0.152666015624999901, 0.023437500000000000
    "Lost In Space", 0, center, -0.746941124267578083, 0.107216993408203093, 0.000048339843750000
    "Lyla's Cross", 0, center, -0.743671323120116967, 0.131868488549804708, 0.000000302734375000
    "Lyla's Flower", 0, center, -1.749537689813964958, -0.000000032808105410, 0.000000409297851611
    "Lyla's Landscape", 0, center, -0.746207083740234145, 0.110563319091796841, 0.000324360351562469
    "Lyla's Place", 0, center, -1.749537680295410347, -0.000000023289550722, 0.000000409297851389
    "Lyla's valley", 0, center, -0.054213312638513089, -0.985251345074024565, 0.000000036428153515
    "The Creepy Blue Vampire Bat", 0, center, -1.427049284697388343, -0.000028839351653103, 0.000000002836980872
    "The Middle of Nowhere", 0, center, -0.054213310436080020, -0.985251353777549577, 0.000000000000396023
    "The spiral of doom", 0, center, -1.749537688871479535, -0.000000098261405626, 0.000000000383716836
    "the undiscovered organism", 0, center, -1.255737840303687847, -0.324752320567795072, 0.000000014248101143
}



old_settings.txt
# This is the settings file for FracGen
# FracGen is (c) Douglas Wolf, 2017


POI =
{
    "Bacterium", ulc, -0.740225588697194681, 0.185894584514200240, 0.000002303123473268
    "Crazy Eyes", ulc, -1.749846063840216459, -0.000000000008744936, 0.000000000002728484
    "Flower 2", ulc, -1.746522467659197497, 0.000000431324283195, 0.000000002561137080
    "Flower 3", ulc, -1.749844146966933911, 0.000000044032931184, 0.000000089406967163
    "Meg's Flower", ulc, -1.746540362918784339, -0.000034233469609375, 0.000000000020000002
    "Seahorse Valley East", ulc, -0.755654296875000075, 0.195556640625000111, 0.023437500000000000
    "Spiderweb", ulc, -0.744973049709637847, 0.209558090424428217, 0.000000000000274558
    "Starfish", ulc, -0.213538556098938098, 0.681681182384490936, 0.000002861022949219
    "Starfish 15", ulc, -0.213537137077655731, 0.681679758292157101, 0.000000002793967724
    "The Garden", ulc, -1.745307589464937692, 0.000000000698464796, 0.000000001396983862
    "The Garden 2", ulc, -1.742122075746731058, 0.000002832412592138, 0.000005722045898438
    "testpint", ulc, -0.743701171875000067, 0.187880859375000098, 0.004716796875000040
}



Settings.cpp
#include "stdafx.h"
#include "Globals.h"
#include "SpecFile.h"
#include "WinUtilsImp.h"

//=========================================================================================================
// Globals local to this file
//=========================================================================================================
static CString  settings_fn = L"settings.txt";
static CSpecFile sf;
//=========================================================================================================

//=========================================================================================================
// ReadSettings() - Reads in the settings file
//=========================================================================================================
bool ReadSettings()
{
    CScript   s;
    poi       place;
   
    // None of these places are built-ins
    place.builtin = false;

    // Open the settings file, and if we can't complain
    if (!sf.Read(settings_fn, false)) return false;

    // If the "POI" spec exists.
    if (sf.Exists(L"poi"))
    {
        // Fetch the script
        sf.Get(L"poi", &s);

        // Loop through each line of the spec...
        while (s.GetNextLine())
        {
            place.name    = s.GetNextWord();
            place.fractal = s.GetNextInt();
            place.center  = s.GetNextWord() == L"center";
            place.real    = s.GetNextFloat();
            place.imag    = s.GetNextFloat();
            place.span    = s.GetNextFloat();
            places[place.name] = place;
        }
    }

    // Tell the caller that all is well
    return true;
}
//=========================================================================================================


//=========================================================================================================
// SaveSettings() - Saves the settings file to the local directory
//=========================================================================================================
bool SaveSettings()
{
    FILE*   ofile;

    // Try to create the file and complain if we can't
    if (_wfopen_s(&ofile, settings_fn, L"w") != 0) return false;

    fprintf(ofile, "# This is the settings file for FracGen\n");
    fprintf(ofile, "# FracGen is (c) Douglas Wolf, 2017\n\n\n");

    // Output a numner that tells us the format of this file
    fprintf(ofile, "FORMAT = 1\n\n");

    // Output the "Points of interest" header
    fprintf(ofile, "POI =\n{\n");

    // Loop through every place we know, ignoring the built-ins
    for (auto it = places.begin(); it != places.end(); ++it)
    {
        // Ignore built-ins
        if (it->second.builtin) continue;
        
        // Fetch the three coordinates
        double real = it->second.real;
        double imag = it->second.imag;
        double span = it->second.span;

        // If they're not 'center' based, make them so
        if (!it->second.center)
        {
            span = span / 2;
            real += span;
            imag -= span;
        }
        
        // Fetch an ASCII version of the name
        CStringA name = toCStringA( it->second.name);

        // Output a record for this point of interest
        fprintf
        (
            ofile,
            "    \"%s\", %u, center, %1.18lf, %1.18lf, %1.18lf\n",
            name.GetBuffer(0),
            it->second.fractal,
            it->second.real,
            it->second.imag,
            it->second.span
        );
    }

    // Tell the caller that we saved his file
    fprintf(ofile, "}\n\n");
    fclose(ofile);
    return true;
}
//=========================================================================================================




//=========================================================================================================
// CreateBuiltinPOI() - Create the builtin list of "places of interest"
//=========================================================================================================
void CreateBuiltinPOI()
{
    poi place;

    // All of the POI's we're about to add are considered "built-in"
    place.builtin = true;
    place.fractal = 0;

    // The default POI
    place.name = NONE_SELECTED;
    places[place.name] = place;

    place.name   = "Psuedo Julia Set";
    place.center = true;
    place.real   = -0.743643;
    place.imag   =  0.131825;
    place.span   = .0002;
    places[place.name] = place;

    place.name   = "Flower and Cauliflower Petals";
    place.center = false;
    place.real   = -1.746540362910034400;
    place.imag   = -0.000034233478359376;
    place.span   = 0.0000000000025;
    places[place.name] = place;

    place.name   = "Meg's Flower";
    place.center = false;
    place.real = -1.746540362918784339;
    place.imag = -0.000034233469609375;
    place.span = 0.000000000020000002;
    places[place.name] = place;

    place.name = "The Garden";
    place.center = false;
    place.real = -1.745307589464937692;
    place.imag = 0.000000000698464796;
    place.span = 0.000000001396983862;
    places[place.name] = place;

    place.name = "The Starfish";
    place.center = true;
    place.real = -0.213537135689672336;
    place.imag = 0.681679756920100743;
    place.span = 0.000045776367187500;
    places[place.name] = place;

    place.name   = "Spacely Sprockets";
    place.center = true;
    place.real   = -0.745390184286497282;
    place.imag   = 0.210221482274606963;
    place.span   = 0.000000070971250534;
    places[place.name] = place;

    place.name   = "Pretty Zoomed-out";
    place.center = true;
    place.real   = -0.742874874023437437;
    place.imag   = 0.131920106582641500;
    place.span   = 0.0128;
    places[place.name] = place;

    place.name   = "The Eye";
    place.center = true;
    place.real   = 0.392526586121475896;
    place.imag   = 0.131765267712362355;
    place.span   = 0.000000000174622983;
    places[place.name] = place;
}
//=========================================================================================================

